tree/ntuple/v7/src/RPageStorageFile.cxx

```diff
@@ -124,49 +126,111 @@ ROOT::Experimental::Internal::RPageSinkFile::CommitSealedPageImpl(DescriptorId_t
    return WriteSealedPage(sealedPage, bytesPacked);
 }
 
+void ROOT::Experimental::Internal::RPageSinkFile::CommitBatchOfPages(CommittedBatch &batch,
+                                                                     std::vector<RNTupleLocator> &locators)
+{
+   Detail::RNTupleAtomicTimer timer(fCounters->fTimeWallWrite, fCounters->fTimeCpuWrite);
+
+   std::uint64_t offset = fWriter->ReserveBlob(batch.fSize, batch.fBytesPacked);
+
+   auto rangeBegin = batch.fBegin.first;
+   auto rangeEnd = batch.fEnd.first;
+   auto firstPageBegin = batch.fBegin.second;
+   auto lastPageEnd = batch.fEnd.second;
+
+   for (auto rangeIt = rangeBegin; rangeIt != rangeEnd; ++rangeIt) {
+      auto pageBegin = (rangeIt == rangeBegin) ? firstPageBegin : rangeIt->fFirst;
+      auto pageEnd = (std::next(rangeIt) == rangeEnd) ? lastPageEnd : rangeIt->fLast;
+
+      locators.reserve(locators.size() + std::distance(pageBegin, pageEnd));
+
+      for (auto sealedPageIt = pageBegin; sealedPageIt != pageEnd; ++sealedPageIt) {
+         fWriter->WriteIntoReservedBlob(sealedPageIt->fBuffer, sealedPageIt->fSize, offset);
+         RNTupleLocator locator;
+         locator.fPosition = offset;
+         locator.fBytesOnStorage = sealedPageIt->fSize;
+         locators.push_back(locator);
+         offset += sealedPageIt->fSize;
+      }
+   }
+
+   fCounters->fNPageCommitted.Add(locators.size());
+   fCounters->fSzWritePayload.Add(batch.fSize);
+   fNBytesCurrentCluster += batch.fSize;
+
+   batch.fSize = 0;
+   batch.fBytesPacked = 0;
+}
+
 std::vector<ROOT::Experimental::RNTupleLocator>
 ROOT::Experimental::Internal::RPageSinkFile::CommitSealedPageVImpl(std::span<RPageStorage::RSealedPageGroup> ranges)
```

Comment:
That is quite sophisticated. Since the PR is anyway already large, can we for now just replace the `bytesPacked >= std::numeric_limits<std::int32_t>::max()` by `bytesPacked >= fOptions->GetMaxKeySize()` and make any further optimization subject to a follow-up PR.